drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+
+Apparently we have an empty state. Not only that, query params disappear from the browser url. What is going on?
+
+Well, most of it is related to Angular's router start up logic. Initially, Angular only emits the actual or real values of the query params through the `activatedRoute.queryParams` observable **when the first navigation** is completed. 
+
+This sounds surprising at first, specially when we try to access a query param that is initially provided in the browser url and, through Angular's router API, its starting value is `undefined`. Why doesn't Angular just pick whatever is initially in the browser url?
+
+To understand it better let's consider an example. Imagine that you visit `http://localhost:4200/home?firstName=Dave` and, in the component assigned to `/home`, you try to access `activatedRoute.snapshot.queryParams["firstName"]`, in its `ngOnInit` for example. Angular won't return `Dave` as you may expect. You will actually receive `undefined`. The reason is that Angular hasn't finished the initial navigation yet. During this period of time, if we access the `snapshot` or the current value of the `activatedRoute.queryParams` observable, we will get an empty object `{}`, that's why we received `undefined` when we try to access an specific query param during start up.
+
+If you think about it, this behavior makes sense, there could be logic configured that affects the initial and all navigations in general, `guards` could be a good example of that. This means, some query params could be added or removed during a navigation, not necessarily matching what was initially available in the browser url. And because of this, Angular is only sure about the actual or correct values of the query params as soon as the current navigation finishes.
+
+This whole situation affects us in the initial value we provide to our `queryParamValue` signal. If we use the `get` method from the `QueryParamService` in components that are loaded during the initial navigation, then our `queryParamValue` signal will be initialized to `undefined`. This is technically not a big deal since it is connected to the `activatedRoute.queryParams` observable. So as soon as Angular emits the actual value of the query param, then the signal will be updated accordingly.
+
+The issue with the empty UI is related to another not so intuitive behavior of Angular's router: Cancelling concurrent navigations. Maybe this surprises you but when we do `router.navigate`, this is actually an asynchronous operation, it returns a `Promise`, which is resolved whenever the navigation is completed.
+
+The problem is when we start or trigger a `router.navigate` when there's a navigation that is still on progress. When Angular detects this, it will **cancel the current navigation** and execute the new one instead, completely ignoring the previous one. How is this related to the empty UI problem? Let's take a look:
```

Comment:
```suggestion
The problem occurs when we trigger `router.navigate` while a navigation is still in progress. When Angular detects this, it will **cancel the current navigation** and execute the new one instead, completely ignoring the previous one. How is this related to the empty UI problem? Let's take a look:
```